americangut/diversity_analysis.py

```diff
@@ -0,0 +1,1729 @@
+
+    _format_axis(ax)
+
+    return ax, xpos
+
+
+def add_comparison_bars(centers, tops, p_values, ax, space=None,
+    interval=None, lowest=None, factor=5, label_size=10,
+    show_value=True):
+    """Adds p_value bars
+
+    The assumes that comparison bars are being introduced for a
+    group of error bars, assuming a reference and comparison.
+
+    Parameters
+    ----------
+    centers : array-like
+        The center of the barplot.
+    tops : array_like
+        The maximum height of the groups to be compared. There must be a
+        top for each center supplied.
+    p_values : array-like
+        The critical values between the groups in the bars. There should be
+        one less p value than there are bars.
+    ax : matplotlib axis, optional
+        The axis where data should be plotted. If none, a new axis instance
+        will be created.
+    space : float
+        The y-distance between each the comparison line.
+    interval : float
+        The space between the bars
+    lowest : float, optional
+        If mutliple sets of comparison bars are being added, this allows
+        the user to set the position of the lowest bar in the group
+    factor : unsigned int, optional
+        The ones-place value to which values will be rounded. For example,
+        for a `factor` of 5, a value of 0.12 will be rounded to 0.15 and a
+        value of 2.7 will be rounded to 3.0.
+    label_size : unsigned int, optional
+        The font size for displaying signifigance labels
+    show_value : bool, optional
+        When True, the signigance bars will display the actual p value.
+        Otherwise, p values will be coded as (p <= 0.1: '+', p <= 0.05: '*',
+        p <= 0.01: '**', p <= 0.001: '***', p <= 0.0001).
+
+    Returns
+    -------
+    lines : list
+        A list of the lines and text objects which have been plotted
+
+    """
+    # Checks the shapes of the inputs
+    if not centers.shape == tops.shape:
+        raise ValueError('centers and tops must be the same length')
+    if not centers.shape[0] == (p_values.shape[0] + 1):
+        raise ValueError('there must be a p-value for each center')
+
+    # Deterines the bar locations
+    if lowest is None:
+        lowest, fudge = _get_bar_height(tops, factor)
+    else:
+        __, fudge = _get_bar_height(tops, factor)
+
+    correct = np.power(10, -np.log10(fudge))
+
+    # Sets the spacing
+    if space is None:
+        space = correct/15.
+    if interval is None:
+        interval = correct/3.
+
+    # Sets the tops of hte bars
+    bar_levels = np.arange(0., len(p_values))*interval + lowest
+
+    # Identifies the center positions for the p text
+    p_cents = [(centers[1] - centers[0])/2. + centers[0]]
+
+    if len(centers) > 2:
+        for idx, center in enumerate(centers[2:]):
+            p_cents.append((center - centers[0])/2. + centers[0])
+    # Determines the p text
+    p_text = []
+    for p in p_values:
```

Comment:
Non-blocking, just a different take on this. The special cases for `show_value` make this difficult to collapse completely unfortunately, though the logic could be split I suppose

```python
    _p_marks = ['', '+', '*', '**', '***', '****', '*****'] # note, 1 index longer than _p_thresh
    _p_thresh = array([1.0, 0.1, 0.05, 0.01, 0.001, 0.0001])
    p_text = []
    for p in p_values:
        if p < 0.01 and show_value:
            p_text.append('%1.e' % p)
        elif show_value:
            p_text.append('%1.2f' % p)
        else:
            p_text.append(_p_marks[np.searchsorted((p > _p_thresh), True)])
```